src/librbd/cache/ReplicatedWriteLogInternal.h

```diff
@@ -0,0 +1,627 @@
+
+template <typename I>
+void ReplicatedWriteLog<I>::rwl_init(Context *on_finish, DeferredContexts &later) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 20) << dendl;
+  TOID(struct WriteLogPoolRoot) pool_root;
+  auto cache_state = dynamic_cast<ImageCacheStateRWL<I>*>(m_cache_state);
+
+  std::lock_guard locker(m_lock);
+  assert(!m_initialized);
+  ldout(cct,5) << "image name: " << m_image_ctx.name << " id: " << m_image_ctx.id << dendl;
+  ldout(cct,5) << "rwl_size: " << cache_state->m_size << dendl;
+  std::string rwl_path = cache_state->m_path;
+  ldout(cct,5) << "rwl_path: " << rwl_path << dendl;
+
+  std::string pool_name = m_image_ctx.md_ctx.get_pool_name();
+  std::string log_pool_name = rwl_path + "/rbd-rwl." + pool_name + "." + m_image_ctx.id + ".pool";
+  std::string log_poolset_name = rwl_path + "/rbd-rwl." + pool_name + "." + m_image_ctx.id + ".poolset";
+  m_log_pool_config_size = max(cache_state->m_size, MIN_POOL_SIZE);
+
+  if (access(log_poolset_name.c_str(), F_OK) == 0) {
+    m_log_pool_name = log_poolset_name;
+    m_log_is_poolset = true;
+  } else {
+    m_log_pool_name = log_pool_name;
+    ldout(cct, 5) << "Poolset file " << log_poolset_name
+		  << " not present (or can't open). Using unreplicated pool" << dendl;
+  }
+
+  if ((!cache_state->m_present) &&
+      (access(m_log_pool_name.c_str(), F_OK) == 0)) {
+    ldout(cct, 5) << "There's an existing pool/poolset file " << m_log_pool_name
+                  << ", While there's no cache in the image metatata." << dendl;
+    if (remove(m_log_pool_name.c_str()) != 0) {
+      lderr(cct) << "Failed to remove the pool/poolset file " << m_log_pool_name
+	         << dendl;
+      on_finish->complete(-errno);
+      return;
+    } else {
+      ldout(cct, 5) << "Removed the existing pool/poolset file." << dendl;
+    }
+  }
+
+  if (access(m_log_pool_name.c_str(), F_OK) != 0) {
+    if ((m_internal->m_log_pool =
+	 pmemobj_create(m_log_pool_name.c_str(),
+			rwl_pool_layout_name,
+			m_log_pool_config_size,
+			(S_IWUSR | S_IRUSR))) == NULL) {
+      lderr(cct) << "failed to create pool (" << m_log_pool_name << ")"
+                   << pmemobj_errormsg() << dendl;
+      m_present = false;
+      m_clean = true;
+      m_empty = true;
+      /* TODO: filter/replace errnos that are meaningless to the caller */
+      on_finish->complete(-errno);
+      return;
+    }
+    m_present = true;
+    m_clean = true;
+    m_empty = true;
+    pool_root = POBJ_ROOT(m_internal->m_log_pool, struct WriteLogPoolRoot);
+
+    /* new pool, calculate and store metadata */
+    size_t effective_pool_size = (size_t)(m_log_pool_config_size * USABLE_SIZE);
+    size_t small_write_size = MIN_WRITE_ALLOC_SIZE + BLOCK_ALLOC_OVERHEAD_BYTES + sizeof(struct WriteLogPmemEntry);
+    uint64_t num_small_writes = (uint64_t)(effective_pool_size / small_write_size);
+    if (num_small_writes > MAX_LOG_ENTRIES) {
+      num_small_writes = MAX_LOG_ENTRIES;
+    }
+    assert(num_small_writes > 2);
```

Comment:
Nit: fail gracefully via `on_finish`

Reply:
Updated with ceph_assert

Reply:
Failed via on_finish. Didn't  use ceph_assert here. Please check this part in ReplicatedWriteLog.cc

Reply:
Done